src/ArrayReader.cpp

```diff
@@ -634,6 +634,29 @@ void AppendArray(shared_ptr<arrow::Array> array_data, K k_array, size_t& index,
   }
 }
 
+void AppendNullBitmap( shared_ptr<arrow::Array> array_data, K k_bitmap, size_t& index )
+{
+    auto type_id = array_data->type_id();
+    if( array_data->null_count() == 0
+        || arrow::Type::LIST == type_id
+        || arrow::Type::LARGE_LIST == type_id
+        || arrow::Type::FIXED_SIZE_LIST == type_id
+        || arrow::Type::MAP == type_id

```

Comment:
This isn't going to work for MAP, STRUCT, UNION, DICT.
In those cases you have to recurse into the the child lists - you're only considering the top level list.
The key thing is that the null bitmap must have the same shape at the values list.
Look at `InitKdbForArray` - you can see how these compound datatype are handled differently - there can be multiple kdb lists for a single arrow array.
These are then populated as special cases by `AppendMap`, `AppendStruct`, `AppendUnion`, `AppendDictionary` which recurse into the child lists.
Take this example - it's a _single_ arrow array:
```
q)\d .arrowkdb
q.arrowkdb)int64_fd:fd.field["int64";dt.int64[]]
q.arrowkdb)float64_fd:fd.field["float64";dt.float64[]]
q.arrowkdb)struct_dt:dt.struct[(int64_fd; float64_fd)]
q.arrowkdb)dt.printDatatype[struct_dt]
struct<int64: int64, float64: double>
q.arrowkdb)ar.prettyPrintArray[struct_dt;((1 2 3);(1.1 2.2 3.3));::]
-- is_valid: all not null
-- child 0 type: int64
  [
    1,
    2,
    3
  ]
-- child 1 type: double
  [
    1.1,
    2.2,
    3.3
  ]
q.arrowkdb)ts.writeReadArray[struct_dt;((1 2 3);(1.1 2.2 3.3));::]
1   2   3
1.1 2.2 3.3
q.arrowkdb)type ts.writeReadArray[struct_dt;((1 2 3);(1.1 2.2 3.3));::]
0h
q.arrowkdb)count ts.writeReadArray[struct_dt;((1 2 3);(1.1 2.2 3.3));::]
2
q.arrowkdb)type each ts.writeReadArray[struct_dt;((1 2 3);(1.1 2.2 3.3));::]
7 9h
q.arrowkdb)
```
The null bitmap has to come back as the same shape - in this case a mixed list of two bool lists.
Then you could support:
```
-- is_valid: all not null
-- child 0 type: int64
  [
    1,
    null,
    3
  ]
-- child 1 type: double
  [
    null,
    2.2,
    3.3
  ]
```